### PR TITLE
Code and documentation style guide

### PR DESCRIPTION
**Pull request by ivan-bocharov**
This is a draft pull request with code and documentation guidelines for FL codebase (`STYLEGUIDE.md`).

While this is still a work in progress I would like to solicit feedback on the current state of this guide.

### CODE REVIEW

**Comment by ThijsvdLaar** [comment]
Very nice work! I have two suggestions:
(1) Consider placing the closing bracket of a long function definition inline:
```
function condense(
    schedule::Schedule, 
    fg::FactorGraph, 
    variable::Symbol, 
    id=generateId(variable))
    ...
end
```
A non-indented bracket on a new line breaks the flow.

(2) Include
```
c = (a + b)*(a - b)
```
under allowed whitespace configurations. Readability depends on the specific formula, and I personally find this configuration more readable in some situations.

**Comment by albertpod** [comment]
Great job @ivan-bocharov!
Agree with @ThijsvdLaar on suggestions. I think it is good to allow a bit more whitespace configurations in style-guide.

**Comment by ivan-bocharov** [comment]
Thanks @ThijsvdLaar @albertpod for your feedback.

I have no strong opinion on (1). I think suggested version also looks nice and readable.
As for (2) - are there any other whitespace configurations that you would like to include? I don't mind including the suggested one.

**Comment by ThijsvdLaar** [comment]
Apparently the Travis build fails for Julia 1 on OSX.
```
┌ Error: Error building `SpecialFunctions`: 
│ gfortran -m64 -O2 -fPIC  -fPIC -c amos/d1mach.f -o amos/d1mach.f.o
│ make: gfortran: No such file or directory
```

**Comment by wmkouw** [comment]
It looks great, Ivan! Cool that we're being so explicit.

My discussion points are:
- Can I tempt you to not recommend one-line functions or methods? I find them more difficult to mentally parse; it takes work to see where the list of arguments stops and where lines end. This is immediately obvious in a multi-line function definition. 
- I'm in favour of the explicit return command.
- The recommendation for breaking lines in method definition lines looks clean. Do we do the same for functions within methods? Or start them at the indentation block? For instance:
```julia
function foo(bar)

    if ..
        for ..
            if ..
                ..
                    X = longFunctionNameWithinMethod(
    arg1::arg1
    arg2::arg2
)
        ..
    end
end
```
or
```julia
function foo(bar)

    if ..
        for ..
            if ..
                ..
                    X = longFunctionNameWithinMethod(
                        arg1::arg1
                        arg2::arg2
                    )
        ..
    end
end
```

**Comment by ivan-bocharov** [comment]
Thanks for your review, @wmkouw 

Regarding discussion points I have following to say:

- While one-line methods are sometimes hard to parse, I firmly believe that we should not outright ban their usage. I think that there are quite a few usecases where using a one-liner results in a readable and not as verbose code. One example from FL codebase that benefits from using one-liners are slug methods for nodes:

  ```julia
  slug(::Type{GaussianMeanPrecision}) = "𝒩"
  ```
  or 
  ```julia
  function slug(::Type{GaussianMeanPrecision})
      return "𝒩"
  end
  ```

  I think that a lot of problematic one-liners fall into a "longer-than-character limit" category. We could also consider some other rules for one-line methods, but I think that most of the biggest offenders in "hard-to-parse" set will be reformatted when we apply new style guidelines to ForneyLab.

- I am in favor of the second variant.

**Comment by ivan-bocharov** [comment]
> Apparently the Travis build fails for Julia 1 on OSX.

The problem is not caused by this PR, `BinaryBuilder.jl` currently doesn't support new version of `gcc` that is default in Travis VMs. The fix is on the way.

**Comment by wmkouw** [comment]
Yeah, good point. Slugs are a nice example where one-liners would be preferable over multi-liners.
